File: packages/mpi/Group.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis
# orthologue
# (c) 1998-2026 all rights reserved
#


# base class
from .Object import Object
# meta-class
from pyre.patterns.Extent import Extent


# declaration
class Group(Object, metaclass=Extent):
    """
    Encapsulation of MPI communicator groups
    """

    # class level public data
    undefined = Object.mpi.undefined


    # per-instance public data
    rank = 0 # my rank in this group
    size = 0 # the size of this group


    # check whether a group is empty
    def isEmpty(self):
        """
        Check whether i am an empty group
        """
        return self.capsule.isEmpty


    # building groups using explicit ranklists
    def include(self, included):
        """
        Build a group out of the processes in {included}
        """
        # build a new group capsule
        capsule = self.capsule.include(list(included))
        # wrap it and return it
        return Group(capsule=capsule)


    def exclude(self, excluded):
        """
        Build a group out of all processes except those in {excluded}
        """
        # build a new group capsule
        capsule = self.capsule.exclude(list(excluded))
        # wrap it and return it
        return Group(capsule=capsule)


    # the set-like operations
    def union(self, g):
        """
        Build a new group whose processes are the union of mine and {g}'s
        """
        return Group(capsule=self.capsule + g.capsule)


    def intersection(self, g):
        """
        Build a new group whose processes are the intersection of mine and {g}'s
        """
        return Group(capsule=self.capsule & g.capsule)


    def difference(self, g):
        """
        Build a new group whose processes are the difference of mine and {g}'s
        """
        return Group(capsule=self.capsule - g.capsule)


    # meta methods
    def __init__(self, capsule, **kwds):
        # chain to my ancestors
        super().__init__(**kwds)

        # store my attributes
        self.capsule = capsule
        # and precompute my rank and size
        self.rank = capsule.rank
        self.size = capsule.size

        # all done
        return


    # implementation details
    capsule = None


# end of file

File: extensions/mpi/Group.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// external dependencies
#include "external.h"
// namespace setup
#include "forward.h"


// add the bindings for the process group
void
pyre::mpi::py::group(py::module & m)
{
    // the class
    auto cls = py::class_<Group>(
        // in scope
        m,
        // the name
        "Group",
        // the docstring
        "a set of processes, ordered so that each one has a rank within the set");

    // the structure
    cls.def_property_readonly(
        // the name
        "rank",
        // the implementation
        &Group::rank,
        // the docstring
        "the rank of this process within me, or {undefined} when it is not a member");

    cls.def_property_readonly(
        // the name
        "size",
        // the implementation
        &Group::size,
        // the docstring
        "the number of processes i hold");

    // whether i hold anybody
    cls.def(
        // the name
        "isEmpty",
        // the implementation
        &Group::isEmpty,
        // the docstring
        "check whether i have no members; note that this asks about membership, not about the "
        "identity of my handle");

    // whether i name a group at all
    cls.def(
        // the name
        "isNull",
        // the implementation
        &Group::isNull,
        // the docstring
        "check whether i name no group at all, as opposed to a group with no members");

    // the truth of a group is whether it names one
    cls.def(
        // the name
        "__bool__",
        // the implementation
        [](const Group & self) -> bool { return static_cast<bool>(self); },
        // the docstring
        "check whether i name a group");

    // building groups out of explicit rank lists
    cls.def(
        // the name
        "include",
        // the implementation
        &Group::include,
        // the signature
        "ranks"_a,
        // the docstring
        "build the group that holds exactly the processes named in {ranks}, in that order");

    cls.def(
        // the name
        "exclude",
        // the implementation
        &Group::exclude,
        // the signature
        "ranks"_a,
        // the docstring
        "build the group that holds all my members except those named in {ranks}");

    // the set operations, which the c++ layer expresses as free functions because neither
    // argument is privileged; python prefers to spell them as methods
    cls.def(
        // the name
        "union",
        // the implementation
        [](const Group & self, const Group & other) -> Group {
            return groupUnion(self, other);
        },
        // the signature
        "other"_a,
        // the docstring
        "build the group whose processes belong to either me or {other}");

    cls.def(
        // the name
        "intersection",
        // the implementation
        [](const Group & self, const Group & other) -> Group {
            return groupIntersection(self, other);
        },
        // the signature
        "other"_a,
        // the docstring
        "build the group whose processes belong to both me and {other}");

    cls.def(
        // the name
        "difference",
        // the implementation
        [](const Group & self, const Group & other) -> Group {
            return groupDifference(self, other);
        },
        // the signature
        "other"_a,
        // the docstring
        "build the group whose processes belong to me but not to {other}");

    // how my membership relates to another's
    cls.def(
        // the name
        "compare",
        // the implementation
        &Group::compare,
        // the signature
        "other"_a,
        // the docstring
        "describe how my membership relates to {other}'s");

    // where my members sit in another group
    cls.def(
        // the name
        "translateRanks",
        // the implementation
        &Group::translateRanks,
        // the signature
        "ranks"_a, "other"_a,
        // the docstring
        "the ranks that my members {ranks} carry in {other}; members that {other} does not hold "
        "come back as {undefined}");

    // for the benefit of anybody staring at a prompt
    cls.def(
        // the name
        "__repr__",
        // the implementation
        [](const Group & self) -> string_t {
            // a null group cannot answer questions about its membership
            if (self.isNull()) {
                return "<mpi.Group: null>";
            }
            // everybody else can
            return "<mpi.Group: rank " + std::to_string(self.rank()) + " of "
                 + std::to_string(self.size()) + ">";
        },
        // the docstring
        "a human readable summary of this group");

    // all done
    return;
}


// end of file
